### PR TITLE
Implementation of break lines, sub & superscripts and bold & italic styles in WebGL texts

### LINKED ISSUE #3196: Line breaks in axis titles of scatter3d
Originally reported in https://github.com/ropensci/plotly/issues/1334

Minimal example -- https://codepen.io/cpsievert/pen/ZqgzEy

### PR DESCRIPTION
**Pull request by archmoj**
This PR fixes https://github.com/ropensci/plotly/issues/1334 and plotly.js #3196 as well as #2551 and #1165 by enabling webgl texts to be drawn on multiple lines using < br > tags. Also another feature is added i.e. to enable using various styles namely italic, bold, sub and super script with webgl texts. 
@etpinard 
@alexcjohnson

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -78,9 +78,9 @@
     "gl-mat4": "^1.2.0",
     "gl-mesh3d": "^2.0.1",
     "gl-plot2d": "^1.3.1",
-    "gl-plot3d": "^1.5.10",
+    "gl-plot3d": "git://github.com/gl-vis/gl-plot3d.git#1c8aadc60eb026bd839f3c4e4a73dab719748843 ",
```
Comment (archmoj) [inline package.json]: https://github.com/mikolalysenko/vectorize-text/compare/master...archmoj:issue-br3d

File: src/plots/gl2d/scene2d.js [hunk 2]
```diff
@@ -20,7 +20,6 @@ var getContext = require('webgl-context');
 
 var createOptions = require('./convert');
 var createCamera = require('./camera');
-var convertHTMLToUnicode = require('../../lib/html2unicode');
```
[the remark this answers is not in the document]
Comment (archmoj) [reply src/plots/gl2d/scene2d.js]: Once I tried removing it failed a few tests. But yes I will give it another try.

File: test/image/mocks/gl3d_multiple-line-texts.json [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        "titlefont": {
+          "color": "#00f",
+          "family": "Times New Roman",
+          "size": 36
+        },
+        "title": "I<sub>x</sub> + z<sup>(2<sub>n</sub>+Log[y<sub>x</sub>])</sup>"
+      },
+      "zaxis":{
+        "titlefont": {
+          "color": "#0f0",
```
[the remark this answers is not in the document]
Comment (archmoj) [reply test/image/mocks/gl3d_multiple-line-texts.json]: Not using different color; but in the mock we used different fonts on the axis which could lock the bug fixed on gl-axes3d.

File: src/plots/gl2d/scene2d.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (archmoj) [reply src/plots/gl2d/scene2d.js]: We were able to remove this and pass all the tests.

[the remark this answers is not in the document]
Comment (etpinard) [reply src/plots/gl2d/scene2d.js]: Cool. I guess now `lib/html2unicode` isn't used anywhere. Could you remove that file altogether?

**Comment by archmoj** [comment]
@etpinard 
RE: <pre>Moreover, I don't understand why gl3d_line_rectangle_render</pre>
<pre>now shows an axis title of the left-hand side. Is this a bug fix?</pre>
Because the axis title was within < > it didn't show up. So yes we may call it a bug fix.

**Comment by etpinard** [comment]
> Because the axis title was within < > it didn't show up. So yes we may call it a bug fix.

Makes sense. Thanks!

-----

Remaining open items:

- [ ] Should dates of 3D axes really be on multiple lines - see gif in https://github.com/plotly/plotly.js/pull/3207#pullrequestreview-172080803
- [ ] Perf testing -> https://codepen.io/etpinard/pen/oQjjgp

**Comment by archmoj** [comment]
Regarding the performance I got these:
<pre>
Previous: 1e4 times "a"
    First Call: 795ms
    Second Call: 485ms
New: 1e4 times "a"
    First Call: 800ms
    Second Call: 475ms

Previous: 1e5 times "a"
    First Call: 3320ms
    Second Call: 2490ms
New: 1e5 times "a"
    First Call: 3200ms
    Second Call: 2500ms

Previous: 1e4 times "H 2 n Bold & Italic"
    First Call: 2900ms
    Second Call: 2000ms

New: 1e4 times "H< sub >2< sup >n< /sup >< /sub >< br >< b >< i >Bold & Italic< /i >< /b >"
    First Call: 2800ms
    Second Call: 1925ms
</pre>

File: package-lock.json [hunk 4]
```diff
@@ -4554,7 +4539,7 @@
         "glslify": "^6.1.0",
         "robust-orientation": "^1.1.3",
         "split-polygon": "^1.0.0",
-        "vectorize-text": "^3.0.0"
+        "vectorize-text": "git://github.com/archmoj/vectorize-text.git#457db88a269e6412c6f6007f1ba88d06bdfbc08c"
```
Comment (archmoj) [inline package-lock.json]: Link to corresponding PR on `vectorize-text` module
https://github.com/mikolalysenko/vectorize-text/pull/8

**Comment by archmoj** [comment]
Thanks for the comment. I could easily control that in `gl-axes3d` and plotly.js side too. 
And I am sure we could keep the `vectorize-text` this way as is more flexible now : )

File: package.json [hunk 5]
```diff
@@ -77,10 +77,10 @@
     "gl-line3d": "^1.1.5",
     "gl-mat4": "^1.2.0",
     "gl-mesh3d": "^2.0.1",
-    "gl-plot2d": "^1.3.1",
-    "gl-plot3d": "^1.5.10",
+    "gl-plot2d": "git://github.com/gl-vis/gl-plot2d.git#bd8a9798589ca4461604ca041479820d6a915103",
```
Comment (archmoj) [inline package.json]: https://github.com/gl-vis/text-cache/compare/issue-br3d

File: package.json [hunk 6]
```diff
@@ -77,10 +77,10 @@
...
+    "gl-plot3d": "git://github.com/gl-vis/gl-plot3d.git#99bd1d20bcec61bdba97cd86bcb2cd5b8b52bd17",
```
Comment (archmoj) [inline package.json]: https://github.com/gl-vis/gl-axes3d/compare/issue-br3d

File: package.json [hunk 7]
```diff
@@ -77,10 +77,10 @@
...
     "gl-pointcloud2d": "^1.0.1",
-    "gl-scatter3d": "^1.0.14",
+    "gl-scatter3d": "git://github.com/gl-vis/gl-scatter3d.git#4353e3493e4e64d363c334fb62ac3e054c50fbf3",
```
Comment (archmoj) [inline package.json]: https://github.com/gl-vis/gl-scatter3d/compare/issue-br3d

**Comment by archmoj** [comment]
The modules are upgraded and the tests are passed on CI. 
@etpinard Could I continue & resolve the conflicts before merge this?

**Comment by etpinard** [comment]
Yes go ahead and resolve the conflict in this PR. Thanks!

File: package-lock.json [hunk 8]
```diff
@@ -4804,7 +4804,7 @@
     },
     "gl-select-static": {
       "version": "2.0.2",
-      "resolved": "https://registry.npmjs.org/gl-select-static/-/gl-select-static-2.0.2.tgz",
+      "resolved": "http://registry.npmjs.org/gl-select-static/-/gl-select-static-2.0.2.tgz",
```
Comment (etpinard) [inline package-lock.json]: Please put back `https` here.

**Comment by etpinard** [comment]
Great PR :dancer:

**Comment by etpinard** [comment]
... make sure to close 

- https://github.com/plotly/plotly.js/issues/3196
- https://github.com/plotly/plotly.js/issues/2551
- https://github.com/plotly/plotly.js/issues/1165

once merged.

**Pull request merged**